File: Assingments/Assingment4/src/com/druid/Huffman/HuffmanNode.java
package com.druid.Huffman;

public class HuffmanNode implements Comparable<HuffmanNode> {
    HuffmanNode left = null;
    HuffmanNode right = null;
    int frequency;
    // This must be initated as null because merged nodes dont necessarily contain individual characters
    Character character = null;

    public HuffmanNode() {
        
    }

    public HuffmanNode(int frequency) {
        this.frequency = frequency;
    }

    public HuffmanNode(int frequency, Character character) {
        this.frequency = frequency;
        this.character = character;
    }

    // We must do this in order to use a minheap with the nodes, as the minheap would have
    // no idea what to compare on if we dont tell it to compare on frequency
    @Override
    public int compareTo(HuffmanNode arg0) {
        if (this.frequency < arg0.frequency) {
            return -1;
        } else if (this.frequency > arg0.frequency) {
            return 1;
        } else {
            return 0;
        }
    }
}

File: Assingments/Assingment4-PythonVersion/Huffman/huffman_node.py
class HuffmanNode:
    def __init__(self, frequency, character, left, right):
        self.frequency = frequency
        self.character = character
        self.left = None
        self.right = None

    # We need this so the min_heap knows what to compare on
    def __lt__(self, other_node):
        return self.frequency < other_node.frequency
